### DISCUSSION TITLE
How to check in the DSL whether a column exists

### DISCUSSION THREAD

**Discussion by halloleo** [johnkerl/miller#1036] [Q&A] [answered]
I have CSVs like this

```
mountain,height,
Everest,8849,
Home Hill,100,My name
```

Now I want to add a better header name for the last column, if it hasn't a good header name yet. I do this with

    mlr --csv -N put 'if(NR==1) { if($[3]=="") { $[3]="comment" } }' file.csv

This works - but only when the 3rd column  exists. An input file like

```
mountain,height
Everest,8849
K2,8611 
```

gives me the error "_mlr: conditional expression did not evaluate to boolean._"

So how can I check that the 3rd column exists before I check for its header value?

**Comment by aborruso** [comment]
It could be useful the `template` verb. Running

```
mlr --csv template -f mountain,height,comment input.csv
```

you will have a good result for both your examples.

**Reply by halloleo** [reply]
Thanks @aborruso! Will try the `template`verb.

However is there a way to check for column existence in the DSL?

**Reply by aborruso** [reply]
> However is there a way to check for column existence in the DSL?

Try this

```
mlr -N --csv put 'if(NR==1) { if(!is_present($3)) { $[3]="comment" } }' then unsparsify input.csv
```

**Reply by halloleo** [reply]
Aah, `is_present`! So cool! Will try both ways later today and report back.

**Reply by halloleo** [reply]
OK, I tried the `template` verb, and it is a great tool to know (to unify the headers in mutiple CSVs), but in my case it doesn't work, beacuse I want to keep the number of columns in each CSV as is.

Then I tried the DSL solution with `is_present` - and it works great! (I didn't use the unparisfy though.) Thanks a lot for your help, @aborruso!

PS: I don't know whether I should mark this answer thread as solution, because the solution evolved for me from the comments not from the initial answer which marking would highlight. - Maybe we leave it as is...

**Comment by aborruso** [comment]
A note. In this

```
mountain,height
Everest,8849
K2,8611
```

the third field does not exist.

Instead here

```
mountain,height,
Everest,8849,
Home Hill,100,My name
```

it exists. it has no label, but it exists.

**Reply by halloleo** [reply]
Yes, indeed that's how the input files can differ.

And the output should reflect this too: The number of existing columns should stay as in the input file.

**Comment by aborruso** [comment, marked as answer]
@halloleo I'm cloning a reply to map a "right" answer.

To check for column existence in the DSL, you could use `is_present` function:

```
mlr -N --csv put 'if(NR==1) { if(!is_present($3)) { $[3]="comment" } }' input.csv
```
